contracts/exchangeIssuance/ExchangeIssuanceLeveraged.sol

```diff
@@ -0,0 +1,1141 @@
+     *
+     * @param _tokens    Addresses of the tokens which need approval
+     */
+    function approveTokens(IERC20[] calldata _tokens) external {
+        for (uint256 i = 0; i < _tokens.length; i++) {
+            _approveToken(_tokens[i]);
+        }
+    }
+
+    /**
+     * Runs all the necessary approval functions required before issuing
+     * or redeeming a SetToken. This function need to be called only once before the first time
+     * this smart contract is used on any particular SetToken.
+     *
+     * @param _setToken    Address of the SetToken being initialized
+     */
+    function approveSetToken(ISetToken _setToken) isSetToken(_setToken) external {
+        LeveragedTokenData memory leveragedTokenData = _getLeveragedTokenData(_setToken, 1 ether, true);
+
+        _approveToken(IERC20(leveragedTokenData.collateralAToken));
+        _approveTokenToLendingPool(IERC20(leveragedTokenData.collateralToken));
+
+        _approveToken(IERC20(leveragedTokenData.debtToken));
+        _approveTokenToLendingPool(IERC20(leveragedTokenData.debtToken));
+    }
+
+    /* ============ Internal Functions ============ */
+
+    /**
+     * Performs all the necessary steps for issuance using the collateral tokens obtained in the flashloan
+     *
+     * @param _collateralToken            Address of the underlying collateral token that was loaned
+     * @param _collateralTokenAmountNet   Amount of collateral token that was received as flashloan
+     * @param _premium                    Premium / Interest that has to be returned to the lending pool on top of the loaned amount
+     * @param _decodedParams              Struct containing token addresses / amounts to perform issuance
+     */
+    function _performIssuance(
+        address _collateralToken,
+        uint256 _collateralTokenAmountNet,
+        uint256 _premium,
+        DecodedParams memory _decodedParams
+    ) 
+    internal 
+    {
+        // Deposit collateral token obtained from flashloan to get the respective aToken position required for issuance
+        _depositCollateralToken(_collateralToken, _collateralTokenAmountNet);
+        // Issue set using the aToken returned by deposit step
+        _issueSet(_decodedParams.setToken, _decodedParams.setAmount, _decodedParams.originalSender);
+        // Obtain necessary collateral tokens to repay flashloan 
+        uint amountInputTokenSpent = _obtainCollateralTokens(
+            _collateralToken,
+            _collateralTokenAmountNet + _premium,
+            _decodedParams.setToken,
+            _decodedParams.setAmount,
+            _decodedParams.originalSender,
+            _decodedParams.exchange,
+            _decodedParams.paymentToken,
+            _decodedParams.limitAmount,
+            _decodedParams.leveragedTokenData
+        );
+        require(amountInputTokenSpent <= _decodedParams.limitAmount, "ExchangeIssuance: INSUFFICIENT INPUT AMOUNT");
```

Comment:
This is where I moved to max/minAmount checks. @ncitron : Let me know if this is "top level" enough.